File: src/lib/slug.ts
/**
 * Slug rules for room links (wolow.app/{slug}).
 *
 * Shared between the client (live validation in the onboarding wizard)
 * and the server (PATCH /api/rooms/[slug]) so the rules can never drift.
 *
 * Product rules: lowercase, 3–20 chars, starts with a letter/number,
 * then letters/numbers/hyphen/underscore. The DB additionally enforces
 * a loose charset/length backstop (migration 027).
 */

export const SLUG_MIN = 3;
export const SLUG_MAX = 20;
export const SLUG_REGEX = /^[a-z0-9][a-z0-9_-]{2,19}$/;

/**
 * Names that collide with app routes / static assets, or that we don't
 * want squatted. Checked against the lowercased candidate.
 */
export const RESERVED_SLUGS = new Set([
  // Existing app routes
  "api",
  "auth",
  "settings",
  "help",
  "sent",
  "welcome",
  "inbox",
  // Static assets / platform paths
  "_next",
  "public",
  "static",
  "assets",
  "icons",
  "images",
  "fonts",
  "sw",
  "manifest",
  "favicon",
  "robots",
  "sitemap",
  // Likely future routes & brand protection
  "about",
  "admin",
  "app",
  "blog",
  "contact",
  "home",
  "login",
  "logout",
  "me",
  "new",
  "official",
  "privacy",
  "root",
  "signin",
  "signout",
  "signup",
  "support",
  "terms",
  "wolow",
]);

export type SlugInvalidReason = "length" | "format" | "reserved";

export type SlugValidation =
  | { ok: true; slug: string }
  | { ok: false; reason: SlugInvalidReason };

export const SLUG_ERROR_MESSAGES: Record<SlugInvalidReason, string> = {
  length: `Must be ${SLUG_MIN}–${SLUG_MAX} characters.`,
  format: "Only lowercase letters, numbers, - and _ (must start with a letter or number).",
  reserved: "That name is reserved — try another one.",
};

/** Normalize + validate a user-supplied slug candidate. */
export function validateSlug(raw: string): SlugValidation {
  const slug = raw.trim().toLowerCase();
  if (slug.length < SLUG_MIN || slug.length > SLUG_MAX) {
    return { ok: false, reason: "length" };
  }
  if (!SLUG_REGEX.test(slug)) {
    return { ok: false, reason: "format" };
  }
  if (RESERVED_SLUGS.has(slug)) {
    return { ok: false, reason: "reserved" };
  }
  return { ok: true, slug };
}

/** Random lowercase alphanumeric suffix, e.g. "x7k". */
export function randomSlugSuffix(length = 3): string {
  const alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";
  let out = "";
  for (let i = 0; i < length; i++) {
    out += alphabet[Math.floor(Math.random() * alphabet.length)];
  }
  return out;
}

/**
 * Suggest a slug from a display name: "Amina Diallo" → "amina".
 * Falls back to padding with random characters when the result is too
 * short, and never returns a reserved name.
 */
export function suggestSlugFromName(name: string): string {
  let base = name
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "") // strip accents
    .replace(/[^a-z0-9]+/g, "") // keep alphanumerics only
    .slice(0, SLUG_MAX);

  if (base.length < SLUG_MIN) {
    base = (base + randomSlugSuffix(SLUG_MIN)).slice(0, SLUG_MAX);
  }
  if (RESERVED_SLUGS.has(base)) {
    base = `${base.slice(0, SLUG_MAX - 3)}${randomSlugSuffix(3)}`;
  }
  return base;
}

File: mobile/lib/core/utils/slug.dart
/// Port of src/lib/slug.ts the rules must never drift from the server,
/// which validates with the same constants in PATCH /api/rooms/[slug].
library;

import 'dart:math';

import 'package:diacritic/diacritic.dart';

const slugMin = 3;
const slugMax = 20;
final slugRegex = RegExp(r'^[a-z0-9][a-z0-9_-]{2,19}$');

/// Names that collide with app routes / static assets, or that we don't
/// want squatted. Checked against the lowercased candidate.
const reservedSlugs = {
  // Existing app routes
  'api', 'auth', 'settings', 'help', 'sent', 'welcome', 'inbox',
  // Static assets / platform paths
  '_next', 'public', 'static', 'assets', 'icons', 'images', 'fonts',
  'sw', 'manifest', 'favicon', 'robots', 'sitemap',
  // API sub-paths under /api/rooms/{slug} (would shadow static routes)
  'provision',
  // Likely future routes & brand protection
  'about', 'admin', 'app', 'blog', 'contact', 'home', 'login', 'logout',
  'me', 'new', 'official', 'privacy', 'root', 'signin', 'signout', 'signup',
  'support', 'terms', 'wolow',
};

enum SlugInvalidReason { length, format, reserved }

const slugErrorMessages = {
  SlugInvalidReason.length: 'Must be $slugMin\u2013$slugMax characters.',
  SlugInvalidReason.format:
      'Only lowercase letters, numbers, - and _ (must start with a letter or number).',
  SlugInvalidReason.reserved: 'That name is reserved try another one.',
};

sealed class SlugValidation {
  const SlugValidation();
}

class SlugValid extends SlugValidation {
  const SlugValid(this.slug);
  final String slug;
}

class SlugInvalid extends SlugValidation {
  const SlugInvalid(this.reason);
  final SlugInvalidReason reason;

  String get message => slugErrorMessages[reason]!;
}

/// Normalize + validate a user-supplied slug candidate.
SlugValidation validateSlug(String raw) {
  final slug = raw.trim().toLowerCase();
  if (slug.length < slugMin || slug.length > slugMax) {
    return const SlugInvalid(SlugInvalidReason.length);
  }
  if (!slugRegex.hasMatch(slug)) {
    return const SlugInvalid(SlugInvalidReason.format);
  }
  if (reservedSlugs.contains(slug)) {
    return const SlugInvalid(SlugInvalidReason.reserved);
  }
  return SlugValid(slug);
}

final _random = Random();

/// Random lowercase alphanumeric suffix, e.g. "x7k".
String randomSlugSuffix([int length = 3]) {
  const alphabet = 'abcdefghijklmnopqrstuvwxyz0123456789';
  final out = StringBuffer();
  for (var i = 0; i < length; i++) {
    out.write(alphabet[_random.nextInt(alphabet.length)]);
  }
  return out.toString();
}

/// Suggest a slug from a display name: "Amina Diallo" -> "amina".
/// Falls back to padding with random characters when the result is too
/// short, and never returns a reserved name.
///
/// Web uses NFKD + combining-mark strip; [removeDiacritics] matches it for
/// the latin scripts that survive the `[^a-z0-9]` filter anyway.
String suggestSlugFromName(String name) {
  var base = removeDiacritics(name.toLowerCase())
      .replaceAll(RegExp(r'[^a-z0-9]+'), '');
  if (base.length > slugMax) base = base.substring(0, slugMax);

  if (base.length < slugMin) {
    base = base + randomSlugSuffix(slugMin);
    if (base.length > slugMax) base = base.substring(0, slugMax);
  }
  if (reservedSlugs.contains(base)) {
    final head = base.length > slugMax - 3 ? base.substring(0, slugMax - 3) : base;
    base = '$head${randomSlugSuffix(3)}';
  }
  return base;
}
